cmd/kubeadm/app/cmd/init.go

```diff
@@ -106,48 +107,71 @@ var (
 		`)))
 )
 
+// initOptions defines all the init options exposed via flags by kubeadm init.
+// Please note that this structure includes the public kubeadm config API, but only a subset of the options
+// supported by this api will be exposed as a flag.
+type initOptions struct {
+	cfgPath               string
+	skipTokenPrint        bool
+	dryRun                bool
+	featureGatesString    string
+	ignorePreflightErrors []string
+	bto                   *options.BootstrapTokenOptions
+	externalcfg           *kubeadmapiv1beta1.InitConfiguration
+}
+
+// initData defines all the runtime information used when running the kubeadm init worklow;
+// this data is shared across all the phases that are included in the workflow.
+type initData struct {
+	cfg                   *kubeadmapi.InitConfiguration
+	skipTokenPrint        bool
+	dryRun                bool
+	ignorePreflightErrors sets.String
+}
+
 // NewCmdInit returns "kubeadm init" command.
 func NewCmdInit(out io.Writer) *cobra.Command {
-	externalcfg := &kubeadmapiv1beta1.InitConfiguration{}
-	kubeadmscheme.Scheme.Default(externalcfg)
-
-	var cfgPath string
-	var skipTokenPrint bool
-	var dryRun bool
-	var featureGatesString string
-	var ignorePreflightErrors []string
-	// Create the options object for the bootstrap token-related flags, and override the default value for .Description
-	bto := options.NewBootstrapTokenOptions()
-	bto.Description = "The default bootstrap token generated by 'kubeadm init'."
+	options := newInitOptions()
+	initRunner := workflow.NewRunner()
 
 	cmd := &cobra.Command{
 		Use:   "init",
 		Short: "Run this command in order to set up the Kubernetes master.",
 		Run: func(cmd *cobra.Command, args []string) {
-			var err error
-			if externalcfg.FeatureGates, err = features.NewFeatureGate(&features.InitFeatureGates, featureGatesString); err != nil {
-				kubeadmutil.CheckErr(err)
-			}
-
-			ignorePreflightErrorsSet, err := validation.ValidateIgnorePreflightErrors(ignorePreflightErrors)
+			c, err := initRunner.InitData()
 			kubeadmutil.CheckErr(err)
 
-			err = validation.ValidateMixedArguments(cmd.Flags())
-			kubeadmutil.CheckErr(err)
+			ctx, _ := c.(initData)
```

Comment:
please rename the ctx variables per our other discussion around confusion.

Reply:
Plus, you can simplify this to `ctx := ...` only.